### PR TITLE
[Fix #306] Add support for reloading code

### LINKED ISSUE #306: Fix :reload-all
### User Story

As a developer, I want to be able to execute a `:reload-all` over any namespace so that I can get all code reloaded to that namespace reloaded.

### Description

Right now when you execute:

* `(require 'commiteth.core)` ( it works)
* `(require 'commiteth.core :reload-all)` (it breaks)

You get this exception:

```
CompilerException clojure.lang.ArityException: Wrong number of args (4) passed to: core/def-sqlvec-fns, compiling:(hugsql/core.clj:342:11) 
```
<details>
<summary>Full Stacktrace</summary>
<pre>
#error{:cause "Wrong number of args (4) passed to: core/def-sqlvec-fns",
       :via [{:type clojure.lang.Compiler$CompilerException,
              :message "clojure.lang.ArityException: Wrong number of args (4) passed to: core/def-sqlvec-fns, compiling:(hugsql/core.clj:342:11)",
              :at [clojure.lang.Compiler analyzeSeq "Compiler.java" 6875]}
             {:type clojure.lang.ArityException,
              :message "Wrong number of args (4) passed to: core/def-sqlvec-fns",
              :at [clojure.lang.Compiler macroexpand1 "Compiler.java" 6781]}],
       :trace [[clojure.lang.Compiler macroexpand1 "Compiler.java" 6781]
               [clojure.lang.Compiler analyzeSeq "Compiler.java" 6854]
               [clojure.lang.Compiler analyze "Compiler.java" 6669]
               [clojure.lang.Compiler analyze "Compiler.java" 6625]
               [clojure.lang.Compiler$BodyExpr$Parser parse "Compiler.java" 6001]
               [clojure.lang.Compiler$FnMethod parse "Compiler.java" 5380]
               [clojure.lang.Compiler$FnExpr parse "Compiler.java" 3972]
               [clojure.lang.Compiler analyzeSeq "Compiler.java" 6866]
               [clojure.lang.Compiler analyze "Compiler.java" 6669]
               [clojure.lang.Compiler analyzeSeq "Compiler.java" 6856]
               [clojure.lang.Compiler analyze "Compiler.java" 6669]
               [clojure.lang.Compiler access$300 "Compiler.java" 38]
               [clojure.lang.Compiler$DefExpr$Parser parse "Compiler.java" 589]
               [clojure.lang.Compiler analyzeSeq "Compiler.java" 6868]
               [clojure.lang.Compiler analyze "Compiler.java" 6669]
               [clojure.lang.Compiler analyze "Compiler.java" 6625]
               [clojure.lang.Compiler eval "Compiler.java" 6931]
               [clojure.lang.Compiler eval "Compiler.java" 6916]
               [clojure.lang.Compiler load "Compiler.java" 7379]
               [clojure.lang.RT loadResourceScript "RT.java" 372]
               [clojure.lang.RT loadResourceScript "RT.java" 363]
               [clojure.lang.RT load "RT.java" 453]
               [clojure.lang.RT load "RT.java" 419]
               [clojure.core$load$fn__5677 invoke "core.clj" 5893]
               [clojure.core$load invokeStatic "core.clj" 5892]
               [clojure.core$load doInvoke "core.clj" 5876]
               [clojure.lang.RestFn invoke "RestFn.java" 408]
               [clojure.core$load_one invokeStatic "core.clj" 5697]
               [clojure.core$load_one invoke "core.clj" 5692]
               [clojure.core$load_lib$fn__5626 invoke "core.clj" 5737]
               [clojure.core$load_lib invokeStatic "core.clj" 5736]
               [clojure.core$load_lib doInvoke "core.clj" 5717]
               [clojure.lang.RestFn applyTo "RestFn.java" 142]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$load_libs invokeStatic "core.clj" 5774]
               [clojure.core$load_libs doInvoke "core.clj" 5758]
               [clojure.lang.RestFn applyTo "RestFn.java" 137]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$require invokeStatic "core.clj" 5796]
               [clojure.core$require doInvoke "core.clj" 5796]
               [clojure.lang.RestFn invoke "RestFn.java" 512]
               [conman.core$eval71157$loading__5569__auto____71158 invoke "core.clj" 1]
               [conman.core$eval71157 invokeStatic "core.clj" 1]
               [conman.core$eval71157 invoke "core.clj" 1]
               [clojure.lang.Compiler eval "Compiler.java" 6927]
               [clojure.lang.Compiler eval "Compiler.java" 6916]
               [clojure.lang.Compiler load "Compiler.java" 7379]
               [clojure.lang.RT loadResourceScript "RT.java" 372]
               [clojure.lang.RT loadResourceScript "RT.java" 363]
               [clojure.lang.RT load "RT.java" 453]
               [clojure.lang.RT load "RT.java" 419]
               [clojure.core$load$fn__5677 invoke "core.clj" 5893]
               [clojure.core$load invokeStatic "core.clj" 5892]
               [clojure.core$load doInvoke "core.clj" 5876]
               [clojure.lang.RestFn invoke "RestFn.java" 408]
               [clojure.core$load_one invokeStatic "core.clj" 5697]
               [clojure.core$load_one invoke "core.clj" 5692]
               [clojure.core$load_lib$fn__5626 invoke "core.clj" 5737]
               [clojure.core$load_lib invokeStatic "core.clj" 5736]
               [clojure.core$load_lib doInvoke "core.clj" 5717]
               [clojure.lang.RestFn applyTo "RestFn.java" 142]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$load_libs invokeStatic "core.clj" 5774]
               [clojure.core$load_libs doInvoke "core.clj" 5758]
               [clojure.lang.RestFn applyTo "RestFn.java" 137]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$require invokeStatic "core.clj" 5796]
               [clojure.core$require doInvoke "core.clj" 5796]
               [clojure.lang.RestFn invoke "RestFn.java" 619]
               [commiteth.db.core$eval70664$loading__5569__auto____70665 invoke "core.clj" 1]
               [commiteth.db.core$eval70664 invokeStatic "core.clj" 1]
               [commiteth.db.core$eval70664 invoke "core.clj" 1]
               [clojure.lang.Compiler eval "Compiler.java" 6927]
               [clojure.lang.Compiler eval "Compiler.java" 6916]
               [clojure.lang.Compiler load "Compiler.java" 7379]
               [clojure.lang.RT loadResourceScript "RT.java" 372]
               [clojure.lang.RT loadResourceScript "RT.java" 363]
               [clojure.lang.RT load "RT.java" 453]
               [clojure.lang.RT load "RT.java" 419]
               [clojure.core$load$fn__5677 invoke "core.clj" 5893]
               [clojure.core$load invokeStatic "core.clj" 5892]
               [clojure.core$load doInvoke "core.clj" 5876]
               [clojure.lang.RestFn invoke "RestFn.java" 408]
               [clojure.core$load_one invokeStatic "core.clj" 5697]
               [clojure.core$load_one invoke "core.clj" 5692]
               [clojure.core$load_lib$fn__5626 invoke "core.clj" 5737]
               [clojure.core$load_lib invokeStatic "core.clj" 5736]
               [clojure.core$load_lib doInvoke "core.clj" 5717]
               [clojure.lang.RestFn applyTo "RestFn.java" 142]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$load_libs invokeStatic "core.clj" 5774]
               [clojure.core$load_libs doInvoke "core.clj" 5758]
               [clojure.lang.RestFn applyTo "RestFn.java" 137]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$require invokeStatic "core.clj" 5796]
               [clojure.core$require doInvoke "core.clj" 5796]
               [clojure.lang.RestFn invoke "RestFn.java" 436]
               [commiteth.db.users$eval70658$loading__5569__auto____70659 invoke "users.clj" 1]
               [commiteth.db.users$eval70658 invokeStatic "users.clj" 1]
               [commiteth.db.users$eval70658 invoke "users.clj" 1]
               [clojure.lang.Compiler eval "Compiler.java" 6927]
               [clojure.lang.Compiler eval "Compiler.java" 6916]
               [clojure.lang.Compiler load "Compiler.java" 7379]
               [clojure.lang.RT loadResourceScript "RT.java" 372]
               [clojure.lang.RT loadResourceScript "RT.java" 363]
               [clojure.lang.RT load "RT.java" 453]
               [clojure.lang.RT load "RT.java" 419]
               [clojure.core$load$fn__5677 invoke "core.clj" 5893]
               [clojure.core$load invokeStatic "core.clj" 5892]
               [clojure.core$load doInvoke "core.clj" 5876]
               [clojure.lang.RestFn invoke "RestFn.java" 408]
               [clojure.core$load_one invokeStatic "core.clj" 5697]
               [clojure.core$load_one invoke "core.clj" 5692]
               [clojure.core$load_lib$fn__5626 invoke "core.clj" 5737]
               [clojure.core$load_lib invokeStatic "core.clj" 5736]
               [clojure.core$load_lib doInvoke "core.clj" 5717]
               [clojure.lang.RestFn applyTo "RestFn.java" 142]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$load_libs invokeStatic "core.clj" 5774]
               [clojure.core$load_libs doInvoke "core.clj" 5758]
               [clojure.lang.RestFn applyTo "RestFn.java" 137]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$require invokeStatic "core.clj" 5796]
               [clojure.core$require doInvoke "core.clj" 5796]
               [clojure.lang.RestFn invoke "RestFn.java" 930]
               [commiteth.routes.redirect$eval70652$loading__5569__auto____70653 invoke "redirect.clj" 1]
               [commiteth.routes.redirect$eval70652 invokeStatic "redirect.clj" 1]
               [commiteth.routes.redirect$eval70652 invoke "redirect.clj" 1]
               [clojure.lang.Compiler eval "Compiler.java" 6927]
               [clojure.lang.Compiler eval "Compiler.java" 6916]
               [clojure.lang.Compiler load "Compiler.java" 7379]
               [clojure.lang.RT loadResourceScript "RT.java" 372]
               [clojure.lang.RT loadResourceScript "RT.java" 363]
               [clojure.lang.RT load "RT.java" 453]
               [clojure.lang.RT load "RT.java" 419]
               [clojure.core$load$fn__5677 invoke "core.clj" 5893]
               [clojure.core$load invokeStatic "core.clj" 5892]
               [clojure.core$load doInvoke "core.clj" 5876]
               [clojure.lang.RestFn invoke "RestFn.java" 408]
               [clojure.core$load_one invokeStatic "core.clj" 5697]
               [clojure.core$load_one invoke "core.clj" 5692]
               [clojure.core$load_lib$fn__5626 invoke "core.clj" 5737]
               [clojure.core$load_lib invokeStatic "core.clj" 5736]
               [clojure.core$load_lib doInvoke "core.clj" 5717]
               [clojure.lang.RestFn applyTo "RestFn.java" 142]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$load_libs invokeStatic "core.clj" 5774]
               [clojure.core$load_libs doInvoke "core.clj" 5758]
               [clojure.lang.RestFn applyTo "RestFn.java" 137]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$require invokeStatic "core.clj" 5796]
               [clojure.core$require doInvoke "core.clj" 5796]
               [clojure.lang.RestFn invoke "RestFn.java" 1289]
               [commiteth.handler$eval60994$loading__5569__auto____60995 invoke "handler.clj" 1]
               [commiteth.handler$eval60994 invokeStatic "handler.clj" 1]
               [commiteth.handler$eval60994 invoke "handler.clj" 1]
               [clojure.lang.Compiler eval "Compiler.java" 6927]
               [clojure.lang.Compiler eval "Compiler.java" 6916]
               [clojure.lang.Compiler load "Compiler.java" 7379]
               [clojure.lang.RT loadResourceScript "RT.java" 372]
               [clojure.lang.RT loadResourceScript "RT.java" 363]
               [clojure.lang.RT load "RT.java" 453]
               [clojure.lang.RT load "RT.java" 419]
               [clojure.core$load$fn__5677 invoke "core.clj" 5893]
               [clojure.core$load invokeStatic "core.clj" 5892]
               [clojure.core$load doInvoke "core.clj" 5876]
               [clojure.lang.RestFn invoke "RestFn.java" 408]
               [clojure.core$load_one invokeStatic "core.clj" 5697]
               [clojure.core$load_one invoke "core.clj" 5692]
               [clojure.core$load_lib$fn__5626 invoke "core.clj" 5737]
               [clojure.core$load_lib invokeStatic "core.clj" 5736]
               [clojure.core$load_lib doInvoke "core.clj" 5717]
               [clojure.lang.RestFn applyTo "RestFn.java" 142]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$load_libs invokeStatic "core.clj" 5774]
               [clojure.core$load_libs doInvoke "core.clj" 5758]
               [clojure.lang.RestFn applyTo "RestFn.java" 137]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$require invokeStatic "core.clj" 5796]
               [clojure.core$require doInvoke "core.clj" 5796]
               [clojure.lang.RestFn invoke "RestFn.java" 805]
               [commiteth.core$eval60988$loading__5569__auto____60989 invoke "core.clj" 1]
               [commiteth.core$eval60988 invokeStatic "core.clj" 1]
               [commiteth.core$eval60988 invoke "core.clj" 1]
               [clojure.lang.Compiler eval "Compiler.java" 6927]
               [clojure.lang.Compiler eval "Compiler.java" 6916]
               [clojure.lang.Compiler load "Compiler.java" 7379]
               [clojure.lang.RT loadResourceScript "RT.java" 372]
               [clojure.lang.RT loadResourceScript "RT.java" 363]
               [clojure.lang.RT load "RT.java" 453]
               [clojure.lang.RT load "RT.java" 419]
               [clojure.core$load$fn__5677 invoke "core.clj" 5893]
               [clojure.core$load invokeStatic "core.clj" 5892]
               [clojure.core$load doInvoke "core.clj" 5876]
               [clojure.lang.RestFn invoke "RestFn.java" 408]
               [clojure.core$load_one invokeStatic "core.clj" 5697]
               [clojure.core$load_all$fn__5618$fn__5621 invoke "core.clj" 5713]
               [clojure.core$load_all$fn__5618 invoke "core.clj" 5713]
               [clojure.lang.AFn call "AFn.java" 18]
               [clojure.lang.LockingTransaction run "LockingTransaction.java" 273]
               [clojure.lang.LockingTransaction runInTransaction "LockingTransaction.java" 229]
               [clojure.core$load_all invokeStatic "core.clj" 5711]
               [clojure.core$load_all invoke "core.clj" 5705]
               [clojure.core$load_lib$fn__5626 invoke "core.clj" 5737]
               [clojure.core$load_lib invokeStatic "core.clj" 5736]
               [clojure.core$load_lib doInvoke "core.clj" 5717]
               [clojure.lang.RestFn applyTo "RestFn.java" 142]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$load_libs invokeStatic "core.clj" 5774]
               [clojure.core$load_libs doInvoke "core.clj" 5758]
               [clojure.lang.RestFn applyTo "RestFn.java" 137]
               [clojure.core$apply invokeStatic "core.clj" 648]
               [clojure.core$require invokeStatic "core.clj" 5796]
               [clojure.core$require doInvoke "core.clj" 5796]
               [clojure.lang.RestFn invoke "RestFn.java" 421]
               [user$eval60984 invokeStatic "form-init8069671980921109939.clj" 1]
               [user$eval60984 invoke "form-init8069671980921109939.clj" 1]
               [clojure.lang.Compiler eval "Compiler.java" 6927]
               [clojure.lang.Compiler eval "Compiler.java" 6890]
               [clojure.core$eval invokeStatic "core.clj" 3105]
               [clojure.core$eval invoke "core.clj" 3101]
               [clojure.main$repl$read_eval_print__7408$fn__7411 invoke "main.clj" 240]
               [clojure.main$repl$read_eval_print__7408 invoke "main.clj" 240]
               [clojure.main$repl$fn__7417 invoke "main.clj" 258]
               [clojure.main$repl invokeStatic "main.clj" 258]
               [clojure.main$repl doInvoke "main.clj" 174]
               [clojure.lang.RestFn invoke "RestFn.java" 1523]
               [clojure.tools.nrepl.middleware.interruptible_eval$evaluate$fn__25412 invoke "interruptible_eval.clj" 87]
               [clojure.lang.AFn applyToHelper "AFn.java" 152]
               [clojure.lang.AFn applyTo "AFn.java" 144]
               [clojure.core$apply invokeStatic "core.clj" 646]
               [clojure.core$with_bindings_STAR_ invokeStatic "core.clj" 1881]
               [clojure.core$with_bindings_STAR_ doInvoke "core.clj" 1881]
               [clojure.lang.RestFn invoke "RestFn.java" 425]
               [clojure.tools.nrepl.middleware.interruptible_eval$evaluate invokeStatic "interruptible_eval.clj" 85]
               [clojure.tools.nrepl.middleware.interruptible_eval$evaluate invoke "interruptible_eval.clj" 55]
               [clojure.tools.nrepl.middleware.interruptible_eval$interruptible_eval$fn__25457$fn__25460
                invoke
                "interruptible_eval.clj"
                222]
               [clojure.tools.nrepl.middleware.interruptible_eval$run_next$fn__25452
                invoke
                "interruptible_eval.clj"
                190]
               [clojure.lang.AFn run "AFn.java" 22]
               [java.util.concurrent.ThreadPoolExecutor runWorker "ThreadPoolExecutor.java" 1142]
               [java.util.concurrent.ThreadPoolExecutor$Worker run "ThreadPoolExecutor.java" 617]
               [java.lang.Thread run "Thread.java" 745]]}
</pre>
</details>

### PR DESCRIPTION
**Pull request by omartell**
Fixes: #306 
Status: Finished

This adds `clojure.tools.namespace` to be able to reload code. It's a better alternative to `reload-all` as outlined [here](https://github.com/clojure/tools.namespace#reloading-code-motivation) and [here](https://github.com/tolitius/mount#the-importance-of-being-reloadable).

If you run:
- `refresh` will reload all changed namespaces with their latest definition.
- `reset` will stop the current system first, then refresh, and restart the
dependencies.

### CODE REVIEW

**Comment by pablodip** [comment]
@omartell Thanks for that!

When you say `refresh`, what fn are you referring to? the ns `user` doesn't have a `refresh` fn.

**Comment by omartell** [comment]
@pablodip it's available on the `user` ns, see https://github.com/omartell/open-bounty/blob/a80c150ffb08af44a90c0f623eb89bf658f1d4e1/env/dev/clj/user.clj#L4.

**Comment by pablodip** [comment]
Got it already, you're referring it from `clojure.tools.namespace.repl`. Could you please create a fn `reload-code` and not refer it from that ns, so that we are explicit about what we can use in the `user` ns?

File: env/dev/clj/user.clj [hunk 1]
```diff
@@ -1,6 +1,9 @@
 (ns user
   (:require [mount.core :as mount]
-            [commiteth.figwheel :refer [start-fw stop-fw cljs]]))
+            [commiteth.figwheel :refer [start-fw stop-fw cljs]]
+            [clojure.tools.namespace.repl :as repl :refer [refresh]]))
+
+(repl/set-refresh-dirs "src" "dev" "test")
```
Comment (pablodip) [inline env/dev/clj/user.clj]: Why exactly is this needed? Isn't the classpath correct?

Comment (omartell) [reply env/dev/clj/user.clj]: In some cases `clojure.tools.namespace` will try to reload code from libraries and break. There are also some issues when reloading cljc files. This prevents that from happening.

Comment (pablodip) [reply env/dev/clj/user.clj]: Could you please explain those two cases (or point to some page that explains them) and how that `set-refresh-dirs` with `"src" "dev" "test"` solves them? Would like to understand it well.

**Comment by omartell** [comment]
@pablodip I think `refresh` is a pretty idiomatic name across Clojure projects, so I think I'll just alias it if you're ok with that. I'll add something like:

```Clojure
(defn refresh []
  (repl/refresh))

```

**Comment by pablodip** [comment]
@omartell Okay with that. :)

**Comment by martinklepsch** [comment]
Maybe we should not encourage using `refresh` directly and instead only provide `reset`? This way we'd minimize variation in the development workflow making it a bit more straightforward to troubleshoot issues.

Also wondering if we really need `restart` and `reset`? Seems like we probably want `reset` most of the time? Guess we should keep `restart` around for a while "just in case" but might be worth considering removing `restart` some time in the future.
https://github.com/omartell/open-bounty/blob/a80c150ffb08af44a90c0f623eb89bf658f1d4e1/env/dev/clj/user.clj#L16-L22

File: env/dev/clj/user.clj (hunk 1 above)
Comment (omartell) [reply env/dev/clj/user.clj]: You can see the issue if you run `clojure.tools.namespace.repl/refresh-all` on the repl a couple of times. More info [here](https://github.com/clojure/tools.namespace#reloading-code-usage) and [here](https://github.com/stuartsierra/component/issues/45#issuecomment-260974812).

**Comment by omartell** [comment]
@martinklepsch @pablodip  I don't think `restart` is really needed if you have `reset`, so can remove if you're ok with that.

File: env/dev/clj/user.clj (hunk 1 above)
Comment (pablodip) [reply env/dev/clj/user.clj]: Okay with it, understood it more, thanks for the links.

**Comment by pablodip** [comment]
@martinklepsch 

> Maybe we should not encourage using refresh directly and instead only provide reset? This way we'd minimize variation in the development workflow making it a bit more straightforward to troubleshoot issues.

`refresh` is a general tool, agnostic of any workflow, so I'd keep it.

@omartell 

> I don't think restart is really needed if you have reset, so can remove if you're ok with that.

I'd keep both and use explanatory names, something like `restart-without-refreshing` and `restart-with-refreshing`.

**Comment by martinklepsch** [comment]
> `refresh` is a general tool, agnostic of any workflow, so I'd keep it.

I don't really see any reasons why we would expose `user/refresh` as part of the `user/` API now. We didn't have it before and the functions in `user/` call it when necessary. Sure it's an agnostic tool but why introduce two things if it's intended that you use `reset` anyways? 

@omartell do you usually call `refresh` on it's own?

**Comment by omartell** [comment]
@martinklepsch yes sometimes, when I don't want the system running but I want the latest definitions, for example when running tests. Spacemacs has a shortcut for it and it's generally available in reloaded  systems when using component.

**Comment by pablodip** [comment]
@martinklepsch 

> I don't really see any reasons why we would expose user/refresh as part of the user/ API now. We didn't have it before and the functions in user/ call it when necessary. Sure it's an agnostic tool but why introduce two things if it's intended that you use reset anyways?
> IMO refresh is an implementation detail that should not be turned into a public API now.

Sometimes you work with parts of the code without starting the whole system.

**Comment by martinklepsch** [comment]
I see. I usually evaluate blocks then but I guess in this context it makes sense to include `refresh` 👍

**Comment by omartell** [comment]
@pablodip @martinklepsch I've updated the PR. I've left `refresh` and `restart` but I've added some doc-strings to make things more clear.

**Comment by martinklepsch** [comment]
@pablodip @churik Do we need to test this separately or should we just merge? It's development code only and @churik seems to be pretty loaded with work already?

**Comment by churik** [comment]
I will merge it by myself, if you are convinced that it is ready to be tested, just move it to `To test` column in `Pipeline for PRs`

**Comment by churik** [comment]
If it has high-prio for testing, let me know

**Comment by pablodip** [comment]
I don't think that this pull requests needs any QA testing, since it only affects local environment.

**Comment by martinklepsch** [comment]
@pablodip merged then. Thanks for contributing @omartell!

**Pull request merged**